### PR TITLE
Kdoc the Kord

### LINKED ISSUE #11: Lots of places lack full documentation
## Summary

Currently the project lacks full documentation with description and informative block tags in lots of places.

## Goal

I open this issue with the intent of starting to fully document the whole project, little by little.

## Motiation

Documentation is key to allow developers to explore the library on their own making it easier and faster to write better code.

## Relevant resources

[Documenting Kotlin Code](https://kotlinlang.org/docs/reference/kotlin-doc.html)
[Dokka](https://github.com/Kotlin/dokka)

/label ~Enhancement 

**Issue originally posted by @Tmpod

### PR DESCRIPTION
**Pull request by NoComment1105**
This PR aims to close #11 or at least tackle the main part of it, by documenting lots and lots of code :D

Anyone who wants to help is welcome to PR to [this branch](https://github.com/NoComment1105/kord/tree/docs-docs-docs) and hopefully we can document as much as possible.

This will remain as a draft for a long time most likely

Any `private` or `internal` classes/functions/variables will not be documented in this PR

What has been documented?
Below is a list of the "parent packages" in each of Kords modules, this is how i categorize the completion of this PR.
**These will be checked once complete**

- [ ] Common
    - [ ] Annotation
    - [ ] Entity
    - [ ] Exception
    - [ ] Ratelimit
    - [ ] Serialization
    - [ ] *Outlying files*

- [ ] Core
    - [ ]  Behavior
    - [x]  Builder
    - [ ]  Cache
    - [ ]  Entity
    - [x] Event
    - [x]  Exception
    - [x]  Gateway
    - [x]  Live
    - [x]  Supplier
    - [x] *Outlying files*

- [ ] Rest
    - [ ] Builder
    - [ ] JSON
    - [ ] Ratelimit
    - [ ] Request
    - [ ] Route
    - [ ] Service
    - [ ] *Outlying files*

- [ ] Voice
    - [ ] Encryption
    - [ ] Exception
    - [ ] Gateway
    - [ ] Handlers
    - [ ] IO
    - [ ] Streams
    - [ ] udp
    - [ ] *Outlying files*

### CODE REVIEW

**Comment by Lukellmann** [comment]
I think for classes that only are a 1:1 mappings of discord json models (like most things in `dev.kord.common.entity`) we shouldn't document every property but instead just link to discord's docs. That means we'll have less work keeping them up to date.

**Comment by NoComment1105** [comment]
> I think for classes that only are 1:1 mappings of discord json models (like most things in `dev.kord.common.entity`) we shouldn't document every property but instead just link to discord's docs. That means we'll have less work keeping them up to date.

Sounds good, I'll change my documenting style to that from now on

File: core/src/main/kotlin/event/channel/ChannelPinsUpdateEvent.kt [hunk 1]
```diff
@@ -1,19 +1,30 @@
...
 import kotlinx.datetime.Instant
 
+/**
+ * The event of a channels Pins being updated.
+ *
+ * @property data The data associated with the event
+ * @property kord The Kord Instance
+ * @property shard The shard for the event
+ * @property customContext Any custom context provided
+ * @property supplier The entity supplier to use for the event
```
Comment (Lukellmann) [inline core/src/main/kotlin/event/channel/ChannelPinsUpdateEvent.kt:26]: these are overridden properties, so the docs should get inherited automatically

File: core/src/main/kotlin/event/channel/ChannelUpdateEvent.kt [hunk 2]
```diff
@@ -45,6 +74,15 @@ public class NewsChannelUpdateEvent(
     }
 }
 
+/**
+ * The event dispatched when a [TextChannel] is updated.
+ *
+ * The [TextChannel] may have been turned into a [NewsChannel].
+ *
+ * The [old] [TextChannel] may be null unless it has been stored in the cache.
```
Comment (Lukellmann) [inline core/src/main/kotlin/event/channel/ChannelUpdateEvent.kt:82]: This will render weird in the dokka output:

The [old](https://kordlib.github.io/kord/core/dev.kord.core.entity.channel/-text-channel/index.html) may be null unless it has been stored in the cache.

You can use this as a workaround:
``The [old][old] [TextChannel] may be `null` unless it has been stored in the cache.``

File: core/src/main/kotlin/event/channel/thread/ThreadListSyncEvent.kt [hunk 3]
```diff
@@ -57,18 +64,36 @@ public class ThreadListSyncEvent(
      */
     public val members: List<ThreadMember> get() = data.members.map { ThreadMember(it, kord) }
 
+    /**
+     * Requests to get the [Guild] for the event
+     *
+     * @throws RequestException if something went wrong while retrieving the guild.
+     * @throws EntityNotFoundException if the guild is null.
+     */
+    // TODO Make this expression body syntax? Maybe another PR?
```
Comment (Lukellmann) [inline core/src/main/kotlin/event/channel/thread/ThreadListSyncEvent.kt:73]: this TODO seems out of scope for this PR

**Comment by Galarzaa90** [comment]
I was recently thinking... there's many undocumented classes, specially `___Data` classes. And I was thinking of doing this (altought probably in various smaller PRs).

And then I found this 😬

**Comment by NoComment1105** [comment]
> I was recently thinking... there's many undocumented classes, specially `___Data` classes. And I was thinking of doing this (altought probably in various smaller PRs).
> 
> And then I found this 😬

Yeah this PR is quite old now, I need to split it down into smaller ones to reduce the size of the PR required. Thanks for reminding me I can do this now :)

**Comment by Galarzaa90** [comment]
> > I was recently thinking... there's many undocumented classes, specially `___Data` classes. And I was thinking of doing this (altought probably in various smaller PRs).
> > And then I found this 😬
> 
> Yeah this PR is quite old now, I need to split it down into smaller ones to reduce the size of the PR required. Thanks for reminding me I can do this now :)

Yeah I was interested in doing this, but I think it would be better if I wait for this to be merged so I could check whatever was left missing (or added later). Glad I decided to do a search before starting 😅

**Comment by NoComment1105** [comment]
> > > I was recently thinking... there's many undocumented classes, specially `___Data` classes. And I was thinking of doing this (altought probably in various smaller PRs).
> > > And then I found this 😬
> > 
> > 
> > Yeah this PR is quite old now, I need to split it down into smaller ones to reduce the size of the PR required. Thanks for reminding me I can do this now :)
> 
> Yeah I was interested in doing this, but I think it would be better if I wait for this to be merged so I could check whatever was left missing (or added later). Glad I decided to do a search before starting 😅

Haha yeah, I'm currently bringing the PR up to date and will then make some smaller PRs to make it easier :)

**Review by Galarzaa90 — COMMENTED** [review]
Some small typos and suggestions.

**Comment by Galarzaa90** [comment]
Damn it, I thought I was looking at the new PR 😭
